src/common/utils/oidc/helper.go

```diff
@@ -186,55 +216,120 @@ func clientCtx(ctx context.Context, verifyCert bool) context.Context {
 	return gooidc.ClientContext(ctx, client)
 }
 
-// RefreshToken refreshes the token passed in parameter, and return the new token.
-func RefreshToken(ctx context.Context, token *Token) (*Token, error) {
-	oauth, err := getOauthConf()
+// refreshToken tries to refresh the token if it's expired, if it doesn't the
+// original one will be returned.
+func refreshToken(ctx context.Context, token *Token) (*Token, error) {
+	oauthCfg, err := getOauthConf()
 	if err != nil {
-		log.Errorf("Failed to get OAuth configuration, error: %v", err)
 		return nil, err
 	}
 	setting := provider.setting.Load().(models.OIDCSetting)
-	ctx = clientCtx(ctx, setting.VerifyCert)
-	ts := oauth.TokenSource(ctx, &token.Token)
-	t, err := ts.Token()
+	cctx := clientCtx(ctx, setting.VerifyCert)
+	ts := oauthCfg.TokenSource(cctx, &token.Token)
+	nt, err := ts.Token()
 	if err != nil {
 		return nil, err
 	}
-
-	it, ok := t.Extra("id_token").(string)
+	it, ok := nt.Extra("id_token").(string)
 	if !ok {
-		log.Debugf("id_token not exist in refresh response")
+		log.Debug("id_token not exist in refresh response")
 	}
-	return &Token{Token: *t, IDToken: it}, nil
+	return &Token{Token: *nt, RawIDToken: it}, nil
 }
 
-// GroupsFromToken returns the list of group name in the token, the claims of the group list is set in OIDCSetting.
-// It's designed not to return errors, in case of unexpected situation it will log and return empty list.
-func GroupsFromToken(token *gooidc.IDToken) []string {
-	if token == nil {
-		log.Warning("Return empty list for nil token")
-		return []string{}
-	}
+// UserInfoFromToken tries to call the UserInfo endpoint of the OIDC provider, and consolidate with ID token
+// to generate a UserInfo object, if the ID token is not in the input token struct, some attributes will be empty
+func UserInfoFromToken(ctx context.Context, token *Token) (*UserInfo, error) {
 	setting := provider.setting.Load().(models.OIDCSetting)
-	if len(setting.GroupsClaim) == 0 {
-		log.Warning("Group claims is not set in OIDC setting returning empty group list.")
-		return []string{}
+	local, err := userInfoFromIDToken(ctx, token, setting)
+	if err != nil {
+		return nil, err
+	}
+	remote, err := userInfoFromRemote(ctx, token, setting)
+	if err != nil {
+		log.Warningf("Failed to get userInfo by calling remote userinfo endpoint, error: %v ", err)
+	}
+	if remote != nil && local != nil {
+		if remote.Subject != local.Subject {
+			return nil, fmt.Errorf("the subject from userinfo: %s does not match the subject from ID token: %s, probably a security attack happened", remote.Subject, local.Subject)
+		}
+		return mergeUserInfo(remote, local), nil
+	} else if remote != nil && local == nil {
+		return remote, nil
+	} else if local != nil && remote == nil {
+		log.Debugf("Fall back to user data from ID token.")
+		return local, nil
+	}
```

Comment:
It is better to put into mergeUserInfo()

Reply:
it is a private func, I don't think it will make any difference